### DISCUSSION TITLE
请问怎么返回socket.onerror的异常到connnect.tsx

### DISCUSSION THREAD

**Discussion by FickXu** [yume-chan/ya-webadb#457] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/7928985/197143245-fcd22f6f-1dfb-407f-a62e-a133e49539a3.png)
依赖库中监听了socket的onerror，在connect.tsx中我不知道怎么去获取返回的异常信息。我试着在依赖库返回一个promise，但在connect.tsx中它的状态为已完成，连接中断的时候（服务器主动断开）无法进入到catch中，请问这种情况该怎么捕获异常？

**Comment by yume-chan** [comment, marked as answer]
Can't understand what you want to do.

When the WebSocket connection is established, the `Promise` will be resolved. A settled `Promise` can't switch to other states anymore.

If server closes the WebSocket connection first, only `close` event will fire, and the listener for that event will close both the `readable` and `writable` streams. `connect.tsx` knows this because the pipe chain will stop and `device.disconnected` will be resolved.

**Reply by FickXu** [reply]
ws在onopen事件触发后直接resolved了，状态不会发生改变。之后服务器主动关闭了ws导致触发了onerror事件和onclose事件，最初我想在依赖包中的onerror中来处理异常信息返回，所以有了上面的疑问。
根据你的解释，我在ws连接后，判断了readable的locked字段的值，如果为false则表示可读流被关闭，从而得知ws已被关闭。请问这个字段可以作为判断依据吗？
这是我的部分代码：
![image](https://user-images.githubusercontent.com/7928985/197445479-70a4cbe7-f309-4e9d-99d3-1d353dbae4f1.png)

**Reply by yume-chan** [reply]
No, the `locked` field indicates whether the stream currently has an active reader (has `getReader` or `pipeTo` called on it). When `connect()` returns, the stream is always available.

Your `onclose` event happens asynchronously, so you also need some asynchronous method to detect it (like a `Promise` or an event). This project generally doesn't use events (because it's easy to be missed, if the listener is attached later than the event is fired). Instead, streams will block if no reader is active, and `Promise` will latch on the state so any time `then()` is called, the same value is returned.

In this situation, `connect()` returns a duplex stream (a pair of readable stream and writable stream). Readable stream will only tell its active reader when it ends, so you can use a `TransformStream` to "intercept" this information, like this:

```ts
const streams = await backend.connect();
const readable = streams.readable.pipeThrough(new TransformStream({
    flush() {
        console.log('Ended');
    }
}));
```

Then pass this readable to `Adb.authenticate()` instead.

---

As said above, the only way to know when a readable stream end is to acquire this reader. In normal data flow, the ADB client will do that, and it has a `disconnected` field, which is a `Promise` to signal connection end. The promise will be resolved when the connection end gracefully and rejected if any error was thrown in the whole lifetime.

connect.tsx already has some code to handle that

https://github.com/yume-chan/ya-webadb/blob/e779f803854a5d2611088f236cf9e21f15f616b0/apps/demo/src/components/connect.tsx#L188-L193

So, your code can also be added here to handle disconnections.

---

However, with any of the above methods, you can only get when the steam ends, but not the error, simply due to WebSocket backend never forwards errors to outside.

This can be done by erroring out the readable stream when error happens, for example by adding this

```ts
socket.onerror = () => {
    controller.error(new Error('WebSocket error')); // or a custom error to wrap error code/message from WS
};
```

after here: 

https://github.com/yume-chan/ya-webadb/blob/6887d8549f27d0c7cf14528f305e1efe2f9b7f4c/libraries/adb-backend-ws/src/index.ts#L36-L39

After that, the `disconnected` promise will become rejected instead of resolved, and the error is what you thrown in backend.

**Reply by FickXu** [reply]
谢谢你耐心解答，我大致明白了连接的逻辑。
虽然在`socket.onerror`中我没有弄明白controller的流是如何产生并返回到页面。但根据你的解释我在`ReadableWritablePair`接口中新增了一个`Promise`类型的类`WSError`用来捕获`socket.onerror`中发生的错误，当它`reject`时，页面`connect.tsx`可以通过`streams.wserror.catch`获取这个异常信息。现在问题已经解决了，谢谢你。
